Reduce RHmin in upper stratosphere and lower mesosphere to be below stratopause RH

DWesl opened the pull request:
Closes #1182 

Alternative to #1181 

Stratospheric specific humidity is mostly set by the specific humidity at the tropopause, with a bit extra from oxidation of methane. When the temperature increases toward the stratopause, the relative humidity falls: I think the saturation vapor pressure winds up less than the total pressure somewhere close to the stratopause, so relative humidity becomes a less useful concept.

The old 1e-6 value led to polar stratopause specific humidity values over 10 mg/kg in the GRIB2 products, around twice the specific humidity found in the rest of the stratosphere. These values do not appear in the model history files.

## 1-degree GRIB2 plots
Monthly-zonal-mean pressure-latitude plots
![2021_q_nontrop_monthly_zonal_means](https://github.com/user-attachments/assets/f0d0df16-3e65-449e-a045-2f1a5df4bfeb)

Area-weighted polar-mean pressure-time plots
![2021_polar_nontrop_q_pressure_time](https://github.com/user-attachments/assets/a4af7ed5-b60b-48a5-82f5-4b8047fb2f85)

## 384-latitude Gaussian-grid history plots
Monthly-zonal-mean pressure-latitude plots
![2021_history_spfh_monthly_zonal_mean_pressure_latitude](https://github.com/user-attachments/assets/789ddfae-06d9-440d-982b-3b4f8962abea)

Area-weighted latitude-band-mean pressure-time plots
![2021_history_spfh_latitude_band_area-weighted_mean_pressure_time (1)](https://github.com/user-attachments/assets/b9d0b0f0-c2fd-482a-bee1-2d10130a2074)

I was asked to provide plots based on the master grib files on a Gaussian grid to ensure the issue is with UPP rather than wgrib2 interpolation, which will take a few hours for data processing.

DWesl committed:
Reduce minimum RH above 3 mbar by a factor of ten.

Old value led to q > 10 ppm near polar stratopause.
This version keeps specific humidity near its values in the rest of the stratosphere.

DWesl committed:
Merge branch 'develop' of NOAA-EMC/UPP into stratopause-rh-lower

WenMeng-NOAA committed:
Merge branch 'develop' into stratopause-rh-lower

WenMeng-NOAA commented:
@DWesl I have run the UPP RTs, and the results appear to show intentional impact only at levels above 3 mb. You may look at my tests at /u/wen.meng/ptmp/UPP-RTs on Cactus.
```
wen.meng@clogin02 UPP-RTs$ ls -ltr */*diff
-rw-r--r-- 1 wen.meng emc  114 Apr 15 16:14 fv3hafs_2022092800/HURPRS09.tm00.diff
-rw-r--r-- 1 wen.meng emc   75 Apr 15 16:16 rap_2020072316_pe_test/WRFPRS.GrbF16.diff
-rw-r--r-- 1 wen.meng emc  114 Apr 15 16:16 fv3hafs_2022092800_pe_test/HURPRS09.tm00.diff
-rw-r--r-- 1 wen.meng emc  791 Apr 15 16:16 fv3r_2023062800/PRSLEV10.tm00.diff
-rw-r--r-- 1 wen.meng emc  791 Apr 15 16:20 fv3r_2023062800_pe_test/PRSLEV10.tm00.diff
-rw-r--r-- 1 wen.meng emc 4301 Apr 15 16:24 fv3gfs_2019083000/gfs.t00z.master.grb2f006.diff
-rw-r--r-- 1 wen.meng emc 4301 Apr 15 16:35 fv3gfs_2019083000_pe_test/gfs.t00z.master.grb2f006.diff
```

DWesl commented:
For the C768 UPP CI test case, maps of specific humidity on available levels between 0.1 and 3 hPa:
The input history file:
<img width="798" alt="2019-08-30T06_history_spfh_stratopause_maps_by_level-or8" src="https://github.com/user-attachments/assets/37d29dc6-aa0e-4094-93ba-89b292013368" />
The expected output:
<img width="616" alt="2019-08-30T06_grib2_control_q_stratopause_maps_by_level-or8" src="https://github.com/user-attachments/assets/feab3a52-5d5f-42a2-82c6-090daaa1296a" />
The output from the CI run (with the changes from this PR):
<img width="616" alt="2019-08-30T06_grib2_experimental_q_stratopause_maps_by_level-or8" src="https://github.com/user-attachments/assets/a0844b01-3def-406e-81dd-189757cf9f91" />

The expected output does show specific humidity values over 10 ppm in the lower mesosphere where the history file and UPP output from this PR show specific humidity below 5 ppm

WenMeng-NOAA commented:
@DWesl Thanks for illustrating the issue and solution using the GFS test case. 
@XiaqiongZhou-NOAA I suggest incorporating this fix, especially since it has minimal impact on non-UFS-based models in UPP.  Please let me your comments.

WenMeng-NOAA commented on sorc/ncep_post.fd/UPP_PHYSICS.f:
```diff
@@ -153,7 +153,11 @@ SUBROUTINE CALRH_NAM(P1,T1,Q1,RH)
-                RH(I,J) = RHmin
+                IF (P1(I, J) >= 3e2) THEN
+                  RH(I,J) = RHmin
+                ELSE IF (RH(I, J) < (RHmin / 10)) THEN
```
@DWesl I suggest changing "RHmin / 10" into "RHmin / 10.".  Also please add a change log entry with your name next to line 119 as "!> 2025-04-16 | ..."

DWesl commented on sorc/ncep_post.fd/UPP_PHYSICS.f:
```diff
@@ -153,7 +153,11 @@ SUBROUTINE CALRH_NAM(P1,T1,Q1,RH)
-                RH(I,J) = RHmin
+                IF (P1(I, J) >= 3e2) THEN
+                  RH(I,J) = RHmin
+                ELSE IF (RH(I, J) < (RHmin / 10)) THEN
+                  RH(I, J) = RHmin / 10
```
Like this?
```suggestion
                ELSE IF (RH(I, J) < (RHmin / 10.)) THEN
                  RH(I, J) = RHmin / 10.
```

WenMeng-NOAA replied:
@DWesl Yes.

DWesl committed:
DOC: Add changelog entry

They may want an upper bound to the change at some point, but I don't know that we get far enough into the mesosphere for that to be an issue

DWesl committed:
BUG: Use floating-point literals to reduce stratopause RHmin

WenMeng-NOAA commented:
@MatthewPyle-NOAA @BenjaminBlake-NOAA @ZhanZhang-NOAA @BinLiu-NOAA @bingfu-NOAA @RuiyuSun  Please be aware of this stratospheric humidity fix in UPP. The impact will be in all UFS-based models if SPFH and RH at above 3 mb levels are outputted from UPP. We will not incorporate this fix in RRFS implementation branch 'release/rrfs_v1' and HAFS implementation branch 'release/hafs_v2'.

WenMeng-NOAA commented:
The UPP RTs have been completed on WCOSS2. There are following changes in gfs, hafs and fv3r tests.
```
gfs:
wen.meng@clogin06 UPP-RTs$ cat fv3gfs_2019083000/gfs.t00z.master.grb2f006.diff
38:88238007:SPFH:0.04 mb:rpn_corr=0.515233:rpn_rms=5.71243e-07
48:119306911:SPFH:0.07 mb:rpn_corr=-0.19218:rpn_rms=2.33695e-06
57:149713039:RH:0.1 mb:rpn_corr=0.999984:rpn_rms=1.66327e-05
58:150899268:SPFH:0.1 mb:rpn_corr=-0.274087:rpn_rms=3.1268e-06
67:179922137:RH:0.2 mb:rpn_corr=0.999476:rpn_rms=1.82261e-05
68:180525564:SPFH:0.2 mb:rpn_corr=-0.142091:rpn_rms=4.02297e-06
77:207762417:RH:0.4 mb:rpn_corr=0.992221:rpn_rms=1.50379e-05
78:208914303:SPFH:0.4 mb:rpn_corr=-0.16802:rpn_rms=2.71568e-06
87:232949881:RH:0.7 mb:rpn_corr=0.992508:rpn_rms=1.0566e-05
88:234088921:SPFH:0.7 mb:rpn_corr=0.221365:rpn_rms=1.22572e-06
97:258118962:RH:1 mb:rpn_corr=0.997286:rpn_rms=7.89276e-06
98:259389490:SPFH:1 mb:rpn_corr=0.526274:rpn_rms=8.37947e-07
107:285535614:RH:2 mb:rpn_corr=0.99996:rpn_rms=6.31555e-06
108:286387248:SPFH:2 mb:rpn_corr=0.845916:rpn_rms=2.92628e-07

hafs:
wen.meng@clogin06 UPP-RTs$ cat fv3hafs_2022092800/HURPRS09.tm00.diff
6:6088348:RH:2 mb:rpn_corr=0.999509:rpn_rms=2.16517e-05
7:6832019:SPFH:2 mb:rpn_corr=0.999007:rpn_rms=5.66475e-08

fv3r:
wen.meng@clogin06 UPP-RTs$ cat fv3r_2023062800_pe_test/PRSLEV10.tm00.diff
49:52099921:RH:2 mb:rpn_corr=0.999966:rpn_rms=1.97957e-06
51:52312694:SPFH:2 mb:rpn_corr=0.9998:rpn_rms=8.71859e-09
```

WenMeng-NOAA commented:
@FernandoAndrade-NOAA This PR is ready for the UPP RTs on R&D machines. You may refer to my testing on Hera at /home/Wen.Meng/stmp2/upp-HERA
```
Wen.Meng@hfe02 upp-HERA$ ls -ltr */*diff
-rw-rw-r--+ 1 Wen.Meng stmp  114 Apr 16 15:48 fv3hafs_2022092800_pe_test/HURPRS09.tm00.diff
-rw-rw-r--+ 1 Wen.Meng stmp  114 Apr 16 15:48 fv3hafs_2022092800/HURPRS09.tm00.diff
-rw-rw-r--+ 1 Wen.Meng stmp   75 Apr 16 15:48 rap_2020072316_pe_test/WRFPRS.GrbF16.diff
-rw-rw-r--+ 1 Wen.Meng stmp  791 Apr 16 15:50 fv3r_2023062800/PRSLEV10.tm00.diff
-rw-rw-r--+ 1 Wen.Meng stmp  791 Apr 16 15:50 fv3r_2023062800_pe_test/PRSLEV10.tm00.diff
-rw-rw-r--+ 1 Wen.Meng stmp 4250 Apr 16 15:56 fv3gfs_2019083000/gfs.t00z.master.grb2f006.diff
-rw-rw-r--+ 1 Wen.Meng stmp 4250 Apr 16 15:56 fv3gfs_2019083000_pe_test/gfs.t00z.master.grb2f006.diff
```

FernandoAndrade-NOAA committed:
add hera RT log

FernandoAndrade-NOAA committed:
add hercules RT log

FernandoAndrade-NOAA committed:
add orion RT log and cmp path fix gfs

FernandoAndrade-NOAA committed:
Update intel.yml update ubuntu version 22.04

FernandoAndrade-NOAA reviewed (approved):
Approved, RT results match expected changes across Hera, Hercules, and Orion. I have added a path fix for the gfs test script's cmp along with a workflow upgrade to ubuntu 22.04 to avoid CI failures. These changes do not affect results.

WenMeng-NOAA commented:
> Approved, RT results match expected changes across Hera, Hercules, and Orion. I have added a path fix for the gfs test script's cmp along with a workflow upgrade to ubuntu 22.04 to avoid CI failures. These changes do not affect results.

@FernandoAndrade-NOAA Thanks for combining those CI-related changes in this PR. They look good to me.

WenMeng-NOAA commented:
This PR is ready for merging.
